mypy/checker.py

```diff
@@ -5058,8 +5168,12 @@ def visit_try_without_finally(self, s: TryStmt, try_frame: bool) -> None:
                             # try/except block.
                             source = var.name
                             if isinstance(var.node, Var):
-                                var.node.type = DeletedType(source=source)
-                            self.binder.cleanse(var)
+                                new_type = DeletedType(source=source)
+                                var.node.type = new_type
+                                if self.options.allow_redefinition_new:
+                                    self.binder.assign_type(var, new_type, new_type)
```

Comment:
Not sure I fully understand the logic here. But again, it seems `.put()` may be used here.

Reply:
Unclear. I will add a TODO comment and investigate it later.